drake/solvers/NloptSolver.cpp

```diff
@@ -0,0 +1,291 @@
+      }
+    }
+  }
+  return result;
+}
+
+void EvaluateVectorConstraint(unsigned m, double* result, unsigned n,
+                              const double* x, double* grad, void* f_data) {
+  const WrappedConstraint* wrapped =
+      reinterpret_cast<WrappedConstraint*>(f_data);
+
+  Eigen::VectorXd xvec(n);
+  for (size_t i = 0; i < n; i++) {
+    xvec[i] = x[i];
+    if (grad) { grad[i] = 0; }
+  }
+
+  const Constraint* c = wrapped->constraint;
+  const size_t num_constraints = c->getNumConstraints();
+  assert(num_constraints == m);
+
+  TaylorVecXd ty(m);
+  TaylorVecXd this_x = MakeInputTaylorVec(xvec, *(wrapped->variable_list));
+  c->eval(this_x, ty);
+
+  const Eigen::VectorXd& lower_bound = c->getLowerBound();
+  const Eigen::VectorXd& upper_bound = c->getUpperBound();
+  for (size_t i = 0; i < num_constraints; i++) {
+    result[i] = ApplyConstraintBounds(
+        ty(i).value(), lower_bound(i), upper_bound(i));
+  }
+
+  if (grad) {
+    for (const DecisionVariableView& v : *(wrapped->variable_list)) {
+      for (size_t i = 0; i < num_constraints; i++) {
+        const double grad_sign =
+            (c->getUpperBound()(i) ==
+             std::numeric_limits<double>::infinity()) ? -1 : 1;
+         for (size_t j = v.index(); j < v.index() + v.size(); j++) {
+           // TODO sam.creasey I don't actually know if summing these
+           // is the right thing to do when more than one constraint
+           // is present.
+           grad[j] += ty(i).derivatives()(j) * grad_sign;
+         }
+      }
+    }
+  }
+}
+}
+
+bool NloptSolver::available() const {
+  return true;
+}
+
+bool NloptSolver::solve(OptimizationProblem &prog) const {
+
+  int nx = prog.getNumVars();
+
+  // Load the algo to use and the size.
+  nlopt::opt opt(nlopt::LD_SLSQP, nx);
+
+  const Eigen::VectorXd& initial_guess = prog.getInitialGuess();
+  std::vector<double> x(initial_guess.size());
+  for (size_t i = 0; i < x.size(); i++) {
+    x[i] = initial_guess[i];
+  }
+
+  std::vector<double> xlow(nx, -std::numeric_limits<double>::infinity());
+  std::vector<double> xupp(nx, std::numeric_limits<double>::infinity());
+
+  for (auto const& binding : prog.getBoundingBoxConstraints()) {
+    auto const& c = binding.getConstraint();
+    for (const DecisionVariableView& v : binding.getVariableList()) {
+      auto const lb = c->getLowerBound(), ub = c->getUpperBound();
+      for (int k = 0; k < v.size(); k++) {
+        const int idx = v.index() + k;
+        xlow[idx] = std::max(lb(k), xlow[idx]);
+        xupp[idx] = std::min(ub(k), xupp[idx]);
+        if (x[idx] < xlow[idx]) { x[idx] = xlow[idx]; }
+        if (x[idx] > xupp[idx]) { x[idx] = xupp[idx]; }
+      }
+    }
+  }
+
+  opt.set_lower_bounds(xlow);
+  opt.set_upper_bounds(xupp);
+
+  opt.set_min_objective(EvaluateCosts, &prog);
+
+  // TODO sam.creasey All hardcoded tolerances in this function should
+  // be made configurable when #1879 is fixed.
+
+  std::list<WrappedConstraint> wrapped_list;
```

Comment:
It actually is to create storage, yes.,  Otherwise the pointers to the WrappedConstraint structs which get passed into NLopt would be invalid since they're created on the stack inside a loop.

I could presumably determine the total number of constraints to wrap ahead of time and allocate them on the stack, but I'm not sure if that's advantageous.